hammertime/rules/timeout.py

```diff
@@ -0,0 +1,90 @@
+# hammertime: A high-volume http fetch library
+# Copyright (C) 2016-  Delve Labs inc.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+
+from time import time
+from statistics import mean, stdev
+
+
+class DynamicTimeout:
+
+    def __init__(self, min_timeout, max_timeout, retries, sample_size=200):
+        self.min_timeout = min_timeout
+        self.max_timeout = max_timeout
+        self.timeout_manager = TimeoutManager(min_timeout, max_timeout, sample_size)
+        self.retries = retries
+        self.request_engine = None
+
+    def set_kb(self, kb):
+        kb.timeout_manager = self.timeout_manager
```

Comment:
The KB should normally only store serializable values. Essentially the results that you could want to transfer.